src/rules/ResourceICUPlurals.js

```diff
@@ -66,98 +66,200 @@ class ResourceICUPlurals extends ResourceRule {
         this.link = "https://gihub.com/ilib-js/i18nlint/blob/main/docs/resource-icu-plurals.md";
     }
 
-    /**
-     * @private
-     */
-    checkPluralCategories(ast, neededCategories, stringToCheck, key, pathName, source, locale) {
-        let value = [];
-        for (let i = 0; i < ast.length; i++) {
-            const opts = ast[i].options;
-            if (opts) {
-                // check if any of the needed categories are missing
-                const missing = neededCategories.filter(category => {
-                    return typeof(opts[category]) === 'undefined';
-                });
-                if ( missing && missing.length ) {
-                    let opts = {
-                        severity: "error",
-                        rule: this,
-                        description: `Missing categories in target string: ${missing.join(", ")}. Expecting these: ${neededCategories.join(", ")}`,
-                        id: key,
-                        highlight: `Target: ${stringToCheck}<e0></e0>`,
-                        pathName,
-                        source,
-                        locale
-                    };
-                    value.push(new Result(opts));
-                }
-                for (let category in opts) {
-                    if ( opts[category] && Array.isArray(opts[category].value) ) {
-                        value = value.concat(this.checkPluralCategories(opts[category].value, neededCategories, stringToCheck, key, pathName, source, locale));
-                    }
-                }
-                // now check the other way around. That is, if the categories that exist are not needed.
-                const extras = Object.keys(opts).filter(category => {
-                    return neededCategories.indexOf(category) < 0;
-                });
-                if (extras && extras.length) {
-                    let opts = {
-                        severity: "warning",
-                        rule: this,
-                        description: `Extra categories in target string: ${extras.join(", ")}. Expecting only these: ${neededCategories.join(", ")}`,
-                        id: key,
-                        highlight: `Target: ${stringToCheck}<e0></e0>`,
-                        pathName,
-                        source,
-                        locale
-                    };
-                    value.push(new Result(opts));
+    matchCategories(sourceSelect, targetSelect, locale, resource) {
+        let problems = [];
+        const srcLocale = new Locale(resource.getSourceLocale());
+
+        // first check the required plural categories
+        let requiredSourceCategories, requiredTargetCategories;
+        let actualRequiredSourceCategories = [], actualRequiredTargetCategories = [];
+        let actualNonrequiredSourceCategories = [], actualNonrequiredTargetCategories = [];
```

Comment:
Could we annotate what are these for near their declaration?
Like
```js
// categories that have to be there according to language rules
let requiredSourceCategories, requiredTargetCategories;
// categories that are in currently processed plural and are required to be there
let actualRequiredSourceCategories = [], actualRequiredTargetCategories = [];
// categories that are in currently processed plural but are not required
let actualNonrequiredSourceCategories = [], actualNonrequiredTargetCategories = [];
```
or just rename them somehow because it's slightly non-obvious what they are